pkg/ipam/node.go

```diff
@@ -782,3 +748,51 @@ func (n *Node) syncToAPIServer() (err error) {
 
 	return err
 }
+
+// update is a helper function for syncToAPIServer(). This function updates the
+// CiliumNode resource spec or status depending on `status`. The resource is
+// updated from `origNode` to `node`.
+func (n *Node) update(
+	origNode, node *v2.CiliumNode,
+	attempts int,
+	status bool,
+) (*v2.CiliumNode, *v2.CiliumNode, error) {
+	scopedLog := n.logger()
+
+	var (
+		updatedNode    *v2.CiliumNode
+		updateErr, err error
+	)
+
+	if status {
+		updatedNode, updateErr = n.manager.k8sAPI.UpdateStatus(origNode, node)
+	} else {
+		updatedNode, updateErr = n.manager.k8sAPI.Update(origNode, node)
+	}
+
+	if updatedNode != nil && updatedNode.Name != "" {
+		node = updatedNode.DeepCopy()
+		if updateErr == nil {
+			return origNode, node, nil
+		}
+	} else if updateErr != nil {
+		scopedLog.WithError(updateErr).WithFields(logrus.Fields{
+			logfields.Attempt: attempts,
+		}).Warning("Failed to update CiliumNode spec")
+
+		node, err = n.manager.k8sAPI.Get(node.Name)
+		if err != nil {
+			return origNode, node, err
```

Comment:
`return nil, nil, err`

Reply:
I don't think we want to do this because this function is called in a loop to retry in case of a failure. If we return nil for `node`, then on the next run, we may try to update the CiliumNode to `nil`. I'll add a comment to clarify that the return values need to stay the same, for this reason.